lib/global-styles-and-settings.php

```diff
@@ -168,6 +168,82 @@ function gutenberg_get_global_styles_custom_css() {
 	return $stylesheet;
 }
 
+/**
+ * Gets the global styles base custom CSS from theme.json.
+ *
+ * @since 6.5.0
+ *
+ * @return string The global base custom CSS.
+ */
+function gutenberg_get_global_styles_base_custom_css() {
+	if ( ! wp_theme_has_theme_json() ) {
+		return '';
+	}
+
+	$can_use_cached = ! WP_DEBUG;
+
+	$cache_key   = 'gutenberg_get_global_styles_base_custom_css';
+	$cache_group = 'theme_json';
+	if ( $can_use_cached ) {
+		$cached = wp_cache_get( $cache_key, $cache_group );
+		if ( $cached ) {
+			return $cached;
+		}
+	}
+
+	$tree       = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data();
+	$stylesheet = $tree->get_base_custom_css();
+
+	if ( $can_use_cached ) {
+		wp_cache_set( $cache_key, $stylesheet, $cache_group );
+	}
+
+	return $stylesheet;
+}
+
+/**
+ * Adds the global styles per-block custom CSS from theme.json
+ * to the inline style for each block.
+ *
+ * @since 6.5.0
+ *
+ *  @global WP_Styles $wp_styles
+ */
+function gutenberg_add_global_styles_block_custom_css() {
+	global $wp_styles;
+
+	if ( ! wp_theme_has_theme_json() || ! wp_should_load_separate_core_block_assets() ) {
+		return '';
+	}
+
+	$tree        = WP_Theme_JSON_Resolver_Gutenberg::get_merged_data();
+	$block_nodes = $tree->get_block_custom_css_nodes();
+
+	foreach ( $block_nodes as $metadata ) {
+		$block_css = $tree->get_block_custom_css( $metadata['css'], $metadata['selector'] );
```

Comment:
Is it worth caching here as well, as is done in `gutenberg_get_global_styles_custom_css()`?

Reply:
I'm not sure, because we'd have to find a way of separately caching the specific combination of blocks on each page. We're not doing it for the global block styles in `gutenberg_add_global_styles_for_blocks`, I assume for the same reason.